File: frontend/apps/web-ele/src/api/thinking/action.ts
import { requestClient } from '#/api/request';

export namespace ThinkingActionApi {
  /** 行动状态 */
  export type ActionStatus = 0 | 1 | 2 | 3; // 0-待开始 1-进行中 2-已完成 3-已取消

  /** 行动信息 */
  export interface ActionInfo {
    id: number;
    title: string;
    description: string;
    userId: number;
    topicId: number;
    topicTitle: string;
    analysisId: number;
    priority: number;
    priorityText: string;
    status: ActionStatus;
    statusText: string;
    progress: number;
    deadline?: string;
    completedAt?: string;
    guidePrinciple: string;
    followUpCount: number;
    isOverdue: boolean;
    createdAt: string;
    updatedAt: string;
  }

  /** 行动详情 */
  export interface ActionDetail extends ActionInfo {}

  /** 创建行动参数 */
  export interface CreateActionParams {
    title: string;
    description?: string;
    topicId?: number;
    topicTitle?: string;
    analysisId?: number;
    priority?: number;
    deadline?: string;
    guidePrinciple?: string;
  }

  /** 更新行动参数 */
  export interface UpdateActionParams {
    id: number;
    title: string;
    description?: string;
    priority?: number;
    deadline?: string;
    guidePrinciple?: string;
  }

  /** 更新行动进度参数 */
  export interface UpdateProgressParams {
    id: number;
    progress: number;
  }

  /** 从分析创建行动参数 */
  export interface CreateFromAnalysisParams {
    analysisId: number;
    topicId?: number;
    topicTitle?: string;
    actions: CreateActionBatch[];
  }

  /** 批量创建行动项 */
  export interface CreateActionBatch {
    title: string;
    description?: string;
    priority?: number;
    deadline?: string;
    guidePrinciple?: string;
  }

  /** 行动列表查询参数 */
  export interface ActionListParams {
    page?: number;
    pageSize?: number;
    title?: string;
    userId?: number;
    topicId?: number;
    analysisId?: number;
    status?: ActionStatus;
    priority?: number;
    isOverdue?: boolean;
  }

  /** 行动列表响应 */
  export interface ActionListResult {
    page: number;
    pageSize: number;
    total: number;
    list: ActionInfo[];
  }

  /** 行动统计 */
  export interface ActionStatistics {
    total: number;
    pendingCount: number;
    progressCount: number;
    completeCount: number;
    cancelCount: number;
    overdueCount: number;
  }

  /** 删除行动参数 */
  export interface DelActionParams {
    ids: number[];
  }
}

/**
 * 获取行动列表
 */
export async function getThinkingActionListApi(params: ThinkingActionApi.ActionListParams = {}) {
  return requestClient.get<ThinkingActionApi.ActionListResult>('/thinking/action/list', { params });
}

/**
 * 获取我的行动列表
 */
export async function getMyThinkingActionListApi(params: ThinkingActionApi.ActionListParams = {}) {
  return requestClient.get<ThinkingActionApi.ActionListResult>('/thinking/action/my', { params });
}

/**
 * 获取行动详情
 */
export async function getThinkingActionDetailApi(id: number) {
  return requestClient.get<ThinkingActionApi.ActionDetail>(`/thinking/action/${id}`);
}

/**
 * 创建行动
 */
export async function createThinkingActionApi(data: ThinkingActionApi.CreateActionParams) {
  return requestClient.post<ThinkingActionApi.ActionInfo>('/thinking/action', data);
}

/**
 * 更新行动
 */
export async function updateThinkingActionApi(data: ThinkingActionApi.UpdateActionParams) {
  return requestClient.put<ThinkingActionApi.ActionInfo>('/thinking/action', data);
}

/**
 * 删除行动
 */
export async function deleteThinkingActionApi(data: ThinkingActionApi.DelActionParams) {
  return requestClient.delete('/thinking/action', { data });
}

/**
 * 从分析创建行动
 */
export async function createActionsFromAnalysisApi(data: ThinkingActionApi.CreateFromAnalysisParams) {
  return requestClient.post<ThinkingActionApi.ActionInfo[]>('/thinking/action/from-analysis', data);
}

/**
 * 获取课题的行动列表
 */
export async function getActionsByTopicApi(topicId: number) {
  return requestClient.get<ThinkingActionApi.ActionInfo[]>(`/thinking/action/by-topic/${topicId}`);
}

/**
 * 获取分析的行动列表
 */
export async function getActionsByAnalysisApi(analysisId: number) {
  return requestClient.get<ThinkingActionApi.ActionInfo[]>(`/thinking/action/by-analysis/${analysisId}`);
}

/**
 * 更新行动进度
 */
export async function updateActionProgressApi(data: ThinkingActionApi.UpdateProgressParams) {
  return requestClient.post('/thinking/action/progress', data);
}

/**
 * 完成行动
 */
export async function completeThinkingActionApi(id: number) {
  return requestClient.post(`/thinking/action/complete/${id}`);
}

/**
 * 取消行动
 */
export async function cancelThinkingActionApi(id: number) {
  return requestClient.post(`/thinking/action/cancel/${id}`);
}

/**
 * 获取行动统计
 */
export async function getThinkingActionStatisticsApi() {
  return requestClient.get<ThinkingActionApi.ActionStatistics>('/thinking/action/statistics');
}

File: backend/logic/practice/action.go
package thinking

import (
	"thinkingModels/domain/practice/action"
	"thinkingModels/logic"

	"github.com/gin-gonic/gin"
)

// ActionLogic 行动项业务逻辑
type ActionLogic struct {
	logic.BaseLogic
}

// NewActionLogic 初始化ActionLogic
func NewActionLogic(ctx *gin.Context) *ActionLogic {
	return &ActionLogic{BaseLogic: logic.BaseLogic{Ctx: ctx}}
}

// Create 创建行动项
func (l *ActionLogic) Create(req *action.CreateAction) (*action.ActionInfo, error) {
	entity := action.NewActionEntity(l.Ctx)
	_, err := entity.SetData(req)
	if err != nil {
		return nil, err
	}

	if concreteEntity, ok := entity.(*action.ActionEntity); ok {
		concreteEntity.Status = 0
		concreteEntity.Progress = 0
	}

	if err := entity.Validate(); err != nil {
		return nil, err
	}

	res, err := entity.Create()
	if err != nil {
		return nil, err
	}

	return convertToActionInfo(res), nil
}

// Update 更新行动项
func (l *ActionLogic) Update(req *action.UpdateAction) (*action.ActionInfo, error) {
	entity := action.NewActionEntity(l.Ctx)
	_, err := entity.LoadById(req.Id)
	if err != nil {
		return nil, err
	}

	_, err = entity.SetData(req)
	if err != nil {
		return nil, err
	}

	if err := entity.Validate(); err != nil {
		return nil, err
	}

	res, err := entity.Update()
	if err != nil {
		return nil, err
	}

	return convertToActionInfo(res), nil
}

// Get 查询行动项详情
func (l *ActionLogic) Get(id uint64) (*action.ActionDetail, error) {
	entity := action.NewActionEntity(l.Ctx)
	res, err := entity.LoadById(id)
	if err != nil {
		return nil, err
	}
	return convertToActionDetail(res), nil
}

// List 查询行动项列表
func (l *ActionLogic) List(req *action.SearchAction) (*logic.ListReap, error) {
	entity := action.NewActionEntity(l.Ctx)
	cond := entity.MakeConditon(*req)

	total, err := entity.Count(cond)
	if err != nil {
		return nil, err
	}

	list, err := entity.List(cond)
	if err != nil {
		return nil, err
	}

	actionInfoList := make([]*action.ActionInfo, 0, len(list))
	for _, item := range list {
		actionInfoList = append(actionInfoList, convertToActionInfo(item))
	}

	return &logic.ListReap{
		Page:     req.Page,
		PageSize: req.PageSize,
		Total:    total,
		List:     actionInfoList,
	}, nil
}

// ListMy 查询我的行动项
func (l *ActionLogic) ListMy(req *action.SearchAction) (*logic.ListReap, error) {
	return l.List(req)
}

// Del 删除行动项
func (l *ActionLogic) Del(req *action.DelAction) error {
	entity := action.NewActionEntity(l.Ctx)
	return entity.Del(req.Ids...)
}

// CreateFromAnalysis 从分析创建行动项
func (l *ActionLogic) CreateFromAnalysis(req *action.CreateActionsFromAnalysis) ([]*action.ActionInfo, error) {
	result := make([]*action.ActionInfo, 0, len(req.Actions))

	for _, actionReq := range req.Actions {
		entity := action.NewActionEntity(l.Ctx)
		_, err := entity.SetData(actionReq)
		if err != nil {
			return nil, err
		}

		if concreteEntity, ok := entity.(*action.ActionEntity); ok {
			concreteEntity.TopicId = req.TopicId
			concreteEntity.AnalysisId = req.AnalysisId
			concreteEntity.Status = 0
			concreteEntity.Progress = 0
		}

		res, err := entity.Create()
		if err != nil {
			return nil, err
		}

		result = append(result, convertToActionInfo(res))
	}

	return result, nil
}

// ListByTopic 按课题查询行动项
func (l *ActionLogic) ListByTopic(topicId uint64) ([]*action.ActionInfo, error) {
	entity := action.NewActionEntity(l.Ctx)
	cond := entity.MakeConditon(action.SearchAction{TopicId: topicId})

	list, err := entity.List(cond)
	if err != nil {
		return nil, err
	}

	result := make([]*action.ActionInfo, 0, len(list))
	for _, item := range list {
		result = append(result, convertToActionInfo(item))
	}

	return result, nil
}

// ListByAnalysis 按分析查询行动项
func (l *ActionLogic) ListByAnalysis(analysisId uint64) ([]*action.ActionInfo, error) {
	entity := action.NewActionEntity(l.Ctx)
	cond := entity.MakeConditon(action.SearchAction{AnalysisId: analysisId})

	list, err := entity.List(cond)
	if err != nil {
		return nil, err
	}

	result := make([]*action.ActionInfo, 0, len(list))
	for _, item := range list {
		result = append(result, convertToActionInfo(item))
	}

	return result, nil
}

// UpdateProgress 更新进度
func (l *ActionLogic) UpdateProgress(req *action.UpdateProgress) error {
	entity := action.NewActionEntity(l.Ctx)
	_, err := entity.LoadById(req.Id)
	if err != nil {
		return err
	}

	if err := entity.UpdateProgress(req.Progress, req.Note); err != nil {
		return err
	}

	_, err = entity.Update()
	return err
}

// Complete 完成行动项
func (l *ActionLogic) Complete(id uint64) error {
	entity := action.NewActionEntity(l.Ctx)
	_, err := entity.LoadById(id)
	if err != nil {
		return err
	}

	if err := entity.MarkComplete(); err != nil {
		return err
	}

	_, err = entity.Update()
	return err
}

// Cancel 取消行动项
func (l *ActionLogic) Cancel(id uint64) error {
	entity := action.NewActionEntity(l.Ctx)
	_, err := entity.LoadById(id)
	if err != nil {
		return err
	}

	if err := entity.Cancel(); err != nil {
		return err
	}

	_, err = entity.Update()
	return err
}

// Statistics 获取统计信息
func (l *ActionLogic) Statistics() (*action.ActionStatistics, error) {
	entity := action.NewActionEntity(l.Ctx)

	pendingCond := entity.MakeConditon(action.SearchAction{Status: intPtrAction(0)})
	pendingCount, _ := entity.Count(pendingCond)

	progressCond := entity.MakeConditon(action.SearchAction{Status: intPtrAction(1)})
	progressCount, _ := entity.Count(progressCond)

	completeCond := entity.MakeConditon(action.SearchAction{Status: intPtrAction(2)})
	completeCount, _ := entity.Count(completeCond)

	cancelCond := entity.MakeConditon(action.SearchAction{Status: intPtrAction(3)})
	cancelCount, _ := entity.Count(cancelCond)

	overdueCond := entity.MakeConditon(action.SearchAction{IsOverdue: boolPtr(true)})
	overdueCount, _ := entity.Count(overdueCond)

	return &action.ActionStatistics{
		PendingCount:  pendingCount,
		ProgressCount: progressCount,
		CompleteCount: completeCount,
		CancelCount:   cancelCount,
		OverdueCount:  overdueCount,
	}, nil
}

func intPtrAction(i int) *int {
	return &i
}

func boolPtr(b bool) *bool {
	return &b
}

func convertToActionInfo(entity any) *action.ActionInfo {
	e, ok := entity.(*action.ActionEntity)
	if !ok {
		return nil
	}
	return &action.ActionInfo{
		Id:            e.Id,
		TopicId:       e.TopicId,
		AnalysisId:    e.AnalysisId,
		Title:         e.Title,
		Description:   e.Description,
		Priority:      e.Priority,
		Status:        e.Status,
		Progress:      e.Progress,
		Deadline:      e.Deadline,
		FollowUpCount: e.FollowUpCount,
		CreatedAt:     e.CreatedAt,
		UpdatedAt:     e.UpdatedAt,
	}
}

func convertToActionDetail(entity any) *action.ActionDetail {
	e, ok := entity.(*action.ActionEntity)
	if !ok {
		return nil
	}
	return &action.ActionDetail{
		Id:             e.Id,
		TopicId:        e.TopicId,
		AnalysisId:     e.AnalysisId,
		Title:          e.Title,
		Description:    e.Description,
		Priority:       e.Priority,
		Status:         e.Status,
		Progress:       e.Progress,
		ExpectedResult: e.ExpectedResult,
		ActualResult:   e.ActualResult,
		Deadline:       e.Deadline,
		CompletedAt:    e.CompletedAt,
		FollowUpCount:  e.FollowUpCount,
		CreatedAt:      e.CreatedAt,
		UpdatedAt:      e.UpdatedAt,
	}
}
